tests/test_pipelineTaskTests.py

```diff
@@ -0,0 +1,416 @@
+    a = connectionTypes.Input(
+        name="PatchA",
+        storageClass="StructuredData",
+        multiple=True,
+        dimensions={"skymap", "tract", "patch"},
+    )
+    b = connectionTypes.Input(
+        name="PatchB",
+        storageClass="StructuredData",
+        multiple=False,
+        dimensions={"skymap", "tract"},
+    )
+    out = connectionTypes.Output(
+        name="PatchOut",
+        storageClass="StructuredData",
+        multiple=True,
+        dimensions={"skymap", "tract", "patch"},
+    )
+
+    def __init__(self, *, config=None):
+        super().__init__(config=config)
+
+        if not config.doUseB:
+            self.inputs.remove("b")
+
+
+class VisitConfig(PipelineTaskConfig, pipelineConnections=VisitConnections):
+    pass
+
+
+class PatchConfig(PipelineTaskConfig, pipelineConnections=PatchConnections):
+    doUseB = lsst.pex.config.Field(default=True, dtype=bool, doc="")
+
+
+class VisitTask(PipelineTask):
+    ConfigClass = VisitConfig
+    _DefaultName = "visit"
+
+    def run(self, a, b):
+        outA = butlerTests.MetricsExample(data=(a.data + b.data))
+        outB = butlerTests.MetricsExample(data=(a.data * max(b.data)))
+        return Struct(outA=outA, outB=outB)
+
+
+class PatchTask(PipelineTask):
+    ConfigClass = PatchConfig
+    _DefaultName = "patch"
+
+    def run(self, a, b=None):
+        if self.config.doUseB:
+            out = [butlerTests.MetricsExample(data=(oneA.data + b.data)) for oneA in a]
+        else:
+            out = a
+        return Struct(out=out)
+
+
+class PipelineTaskTestSuite(lsst.utils.tests.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        # Repository should be re-created for each test case, but
+        # this has a prohibitive run-time cost at present
+        cls.root = tempfile.mkdtemp()
+
+        dataIds = {
+            "instrument": ["notACam"],
+            "physical_filter": ["k2020"],  # needed for expandUniqueId(visit)
+            "visit": [101, 102],
+            "skymap": ["sky"],
+            "tract": [42],
+            "patch": [0, 1],
+        }
+        cls.repo = butlerTests.makeTestRepo(cls.root, dataIds)
+        butlerTests.registerMetricsExample(cls.repo)
+
+        for typeName in {"VisitA", "VisitB", "VisitOutA", "VisitOutB"}:
+            butlerTests.addDatasetType(cls.repo, typeName, {"instrument", "visit"}, "StructuredData")
+        for typeName in {"PatchA", "PatchOut"}:
+            butlerTests.addDatasetType(cls.repo, typeName, {"skymap", "tract", "patch"}, "StructuredData")
+        butlerTests.addDatasetType(cls.repo, "PatchB", {"skymap", "tract"}, "StructuredData")
+
+    @classmethod
+    def tearDownClass(cls):
+        shutil.rmtree(cls.root, ignore_errors=True)
+        super().tearDownClass()
+
+    def setUp(self):
+        super().setUp()
+        self.butler = butlerTests.makeTestCollection(self.repo)
+
+    def testMakeQuantumNoSuchDatatype(self):
```

Comment:
My instinct in these tests is to have significantly fewer functions with multiple situations tested in each so you don't have to duplicate the setup code lines as much. For instance, `testMakeQuantum` which would have a comment indicating that where you test the "no such dataset type" case, then later a comment that you're testing the "invalid dimension" case, then the "missing multiple" case, etc. Is there a stylistic or other reason you prefer to do it this way?

Reply:
It's considered good practice to have a separate function for each test (testing purists would say any function with multiple asserts is too big, though I wouldn't go that far). It means you get more information up-front about what is working and what is failing, and you don't have to worry (in theory) about different cases interfering with each other.

I can try to factor the setup code into separate functions. The main problem is that in a few cases (like the "WithRun" tests) the test really does depend on the details of the mock data.

Reply:
That's fair, thanks for the context about test good practice. I don't want you to spend too much time on this, but if you can factor out a bit more of the setup code (e.g., so you don't define `inA = butlerTests.MetricsExample(data=[1, 2, 3])` multiple times), that would make it easier to read.